### PR TITLE
Add solutions and modify `README.md` for lab01

### PR DESCRIPTION
**Pull request by RCX777**
Add solutions for the exercises to the _solution/_ directory.

The `README.md` is hard to read at times, mainly because of crowded paragraphs and values such as `0x564c364482a0`, positioned inline with other text, and without being **emphasized** with formatting. Update the formatting of the file to allow for better visibility.

Add a small example at the beginning of the `README.md` for variable prefixes in C, such as `0x` or `0b`.

### CODE REVIEW

File: laborator/content/reprezentare-numere/README.md
Comment (teodutu) [inline laborator/content/reprezentare-numere/README.md:250]:
```suggestion
> ```
> make build
>
> make run
> ```
```

File: laborator/content/reprezentare-numere/README.md
Comment (teodutu) [inline laborator/content/reprezentare-numere/README.md:343]: I would honestly remove this mention altogether. It's obvious what the `solution/` folder contains and writing it down might make students overly prone to consult the solution without trying to solve the exercises on their own first.

File: laborator/content/reprezentare-numere/solution/2-len_xor/Makefile
Comment (teodutu) [inline laborator/content/reprezentare-numere/solution/2-len_xor/Makefile:2]: Use the makefiles from the `utils/` folder currently present in the repo: https://github.com/systems-cs-pub-ro/iocla/tree/master/laborator/content/utils.

File: laborator/content/reprezentare-numere/solution/utils/Makefile.generic
Comment (teodutu) [inline laborator/content/reprezentare-numere/solution/utils/Makefile.generic:1]: All these files are already present in the repo. No need to duplicate them here.

**Review by teodutu — CHANGES_REQUESTED** [review]
The changes to the documentation are mostly good and the solutions are alright as well. Fix my inline comments and then this is good to go.

File: laborator/content/reprezentare-numere/solution/2-len_xor/Makefile
Comment (teodutu) [inline laborator/content/reprezentare-numere/solution/2-len_xor/Makefile:1]: On second thought, it would be better to just change this file to a link to `../../2-len_xor/Makefile`. This will make this makefile smaller and any changes to the skeleton makefile will automatically be replicated to the solution without losing functionality. Do the same for all the other makefiles in the `solution/` folder.

File: laborator/content/reprezentare-numere/solution/5-odd_even/odd_even.c
Comment (teodutu) [inline laborator/content/reprezentare-numere/solution/5-odd_even/odd_even.c:11]:
```suggestion
	bits = malloc(sizeof(*bits) * nr_bits);
```
It's recommended that you use `sizeof(variable)` instead of `sizeof(type)` so the size is updated automatically if you change the type of the variable.

File: laborator/content/reprezentare-numere/solution/5-odd_even/odd_even.c
Comment (teodutu) [inline laborator/content/reprezentare-numere/solution/5-odd_even/odd_even.c:59]:
```suggestion
	int *numbers, i, 

	printf("Size of array: ");
	scanf("%d", &n);

	numbers = malloc(sizeof(*numbers) * n);

	for (i = 0; i < n; ++i) {
		printf("Number %d: ", i + 1);
		scanf("%d", numbers + i);
	}

	check_parity(numbers, n);

	// free memory
```
Why allocate 4 bytes? Just use a local variable. And ditto about `sizeof(*numbers)`. Also add some error handling in case that `malloc()` fails.

**Review by teodutu — CHANGES_REQUESTED** [review]
I think it would be better to change the makefiles in the `solution/` folder to links to the ones in the skeletons. And remove the unnecessary allocation in `odd_even.c`.

File: laborator/content/reprezentare-numere/solution/2-len_xor/Makefile
Comment (RCX777) [reply laborator/content/reprezentare-numere/solution/2-len_xor/Makefile:1]: Would this mean that after the required modifications, the Makefiles in the `solution/` directory will use the ones in the lab skeleton, which in turn use the generic in the common `utils/` directory, or am I getting this wrong?

File: laborator/content/reprezentare-numere/solution/5-odd_even/odd_even.c
Comment (RCX777) [reply laborator/content/reprezentare-numere/solution/5-odd_even/odd_even.c:59]: I have a Github-related question here. I can directly press a commit button and your modifications will be commited to this PR, but the message would end in the line "Co-authored by: ...", which is not recognized by the `checkpatch` script, and will result in the PR failing the automated tests. This happened before, and I had to squash the commits and force-push. Is there a way I can change the commit message to allow for a "Co-authored by:" statement?

Comment (teodutu) [reply laborator/content/reprezentare-numere/solution/5-odd_even/odd_even.c:59]: Yeah `checkpatch.pl` checks for `Co-developed by:` [1] :upside_down_face:. But you can change the commit message when you apply suggestions. Follow step 4 from here [2].

[1] https://github.com/torvalds/linux/blob/ef5f68cc1f829b492b19cd4df5af4454aa816b93/scripts/checkpatch.pl#L613
[2] https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/incorporating-feedback-in-your-pull-request#applying-suggested-changes

File: laborator/content/reprezentare-numere/solution/2-len_xor/Makefile
Comment (teodutu) [reply laborator/content/reprezentare-numere/solution/2-len_xor/Makefile:1]: Yup, you're getting this right. The makefiles for the solutions will simply be symbolic links pointing to the ones from the skeletons, similar to how the lab folders [1] are links to subfolders in the `content/` directory [1].

[1] https://github.com/systems-cs-pub-ro/iocla/tree/master/laborator

Comment (RCX777) [reply laborator/content/reprezentare-numere/solution/2-len_xor/Makefile:1]: There seems to be a problem with this approach. While trying to implement this change, I get greeted with the following error message:

image

This happens both when using soft-links or hard-links using the `ln` or `ln -s` command.

When calling `make`, the program will parse the contents of the file pointed to by the symlink, which specify that the `Makefile.generic` should be located at `../../utils/Makefile.generic`, relative to the `Makefile` in `lab/exercise/Makefile`, but when calling `make` from `lab/solution/exercise/Makefile` (symlink), `make` will think that this is the current path, and not `lab/exercise/Makefile`.

Is there a workaround to this problem ?

Comment (teodutu) [reply laborator/content/reprezentare-numere/solution/2-len_xor/Makefile:1]: The simplest solution would be to move the skeleton code in a `support` folder so that the paths from all Makefiles to the global ones are the same. @darius-m do you have a better one?

Comment (darius-m) [reply laborator/content/reprezentare-numere/solution/2-len_xor/Makefile:1]: If you want to add a `support` or `skel` directory for consistency with `sol`, I think that could be a good idea. But changing the directory structure just to be able to create symlinks to makefiles that contain two or three lines of variable definitions and includes is probably more work than it's worth, since updating just the makefile (without changing the task itself) is likely an uncommon occurance.

Comment (teodutu) [reply laborator/content/reprezentare-numere/solution/2-len_xor/Makefile:1]: You're right, @darius-m, that changing the structure for just a few lines that we are going to remove next semester when we move to OEH, is redundant. Before we merge this, I'll ask around first. Maybe there's a simpler solution.

Comment (teodutu) [reply laborator/content/reprezentare-numere/solution/2-len_xor/Makefile:1]: @RCX777 yeah it seems that using links is going to be too much of a hustle. Let's keep the current format.

**Review by teodutu — APPROVED** [review]

**Pull request merged**
